lib/ansible/plugins/inventory/__init__.py

```diff
@@ -316,6 +316,7 @@ def _add_host_to_composed_groups(self, groups, variables, host, strict=False):
         ''' helper to create complex groups for plugins based on jinja2 conditionals, hosts that meet the conditional are added to group'''
         # process each 'group entry'
         if groups and isinstance(groups, dict):
+            variables.update(self.inventory.get_host(host).vars)
```

Comment:
you might want to use combine vars  function, in case people want to control 'hash behaviour'

Reply:
That's a good thought, done.